contracts/minter-imago/src/contract.rs

```diff
@@ -0,0 +1,743 @@
+use cosmwasm_std::{Addr, BankMsg, Binary, coin, Coin, coins, CosmosMsg, Decimal, Deps, DepsMut, Empty, Env, MessageInfo, Order, Reply, ReplyOn, StdError, StdResult, Timestamp, to_binary, Uint128, WasmMsg};
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cw2::set_contract_version;
+use cw721_base::{MintMsg, msg::ExecuteMsg as Cw721ExecuteMsg};
+use cw_utils::{may_pay, maybe_addr, must_pay, parse_reply_instantiate_data};
+use sg1::{checked_fair_burn, FeeError};
+use sg_std::{GENESIS_MINT_START_TIME, NATIVE_DENOM, StargazeMsgWrapper};
+use url::Url;
+
+use sg721_imago::msg::InstantiateMsg as Sg721InstantiateMsg;
+use whitelist::msg::{
+    ConfigResponse as WhitelistConfigResponse, HasMemberResponse, QueryMsg as WhitelistQueryMsg,
+};
+
+use crate::error::ContractError;
+use crate::msg::{
+    ConfigResponse, ExecuteMsg, InstantiateMsg, MintableNumTokensResponse, MintCountResponse,
+    MintPriceResponse, QueryMsg, StartTimeResponse,
+};
+use crate::state::{
+    Config, CONFIG, MINTABLE_NUM_TOKENS, MINTABLE_TOKEN_IDS, MINTER_ADDRS, MINTER_LAST_BLOCK, SG721_ADDRESS,
+};
+
+pub type Response = cosmwasm_std::Response<StargazeMsgWrapper>;
+pub type SubMsg = cosmwasm_std::SubMsg<StargazeMsgWrapper>;
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:sg-minter-imago";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+const DEV_ADDRESS: &str = "stars1zmqesn4d0gjwhcp2f0j3ptc2agqjcqmuadl6cr";
+
+const PW_HOSTNAME_SUFFIX: &str = "publicworks.art";
+
+const INSTANTIATE_SG721_REPLY_ID: u64 = 1;
+
+// governance parameters
+const MAX_TOKEN_LIMIT: u32 = 10000;
```

Comment:
would each project on publicworks have the same limit?

Reply:
It's configurable by the creator, like any other nft collection. It doesn't need a max but this is a reasonable limit I kept from the original contract.